### PR TITLE
TMDM-13276 [Impact Analysis] ElementType change (length) not reflected in PostgreSQL database tables

### PR DESCRIPTION
**Pull request by zfkang-talend**
Jira: https://jira.talendforge.org/browse/TMDM-13276

**What is the current behavior?** (You should also link to an open issue here)
1. If increase the length for one string field, or db2, if greater than 255, for Oracle, if greater than 4000,  will redeploy failure.

**What is the new behavior?**
1. if increase the length for one string field, for db2, if greater than 255, for Oracle, if greater than 4000, will give high impact result, 

**Please check if the PR fulfills these requirements**

- [x] The commit message follows Talend standard
- [x] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features) ?

**What kind of change does this PR introduce?**

- [x] Bugfix
- [ ] Feature
- [ ] Code style update (formatting, local variables)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build / CI related changes
- [ ] Other... Please describe:

**Does this PR introduce a breaking change?**

- [ ] Yes
- [x] No

If this PR contains a breaking change, please describe the impact and migration path for existing applications: ...

**Other information**:

### CODE REVIEW

File: main/plugins/org.talend.mdm.commmon/src/main/java/org/talend/mdm/commmon/metadata/compare/Change.java [hunk 1]
```diff
@@ -43,4 +51,12 @@ public boolean equals(Object o) {
...
         return 31 * this.element.hashCode();
     }
+
+    public Map<String, Object> getData() {
+        return data;
+    }
+
+    public void addData(String key, Object value){
+        this.data.put(key, value);
+    }
```
Comment (hw-zhu) [inline main/plugins/org.talend.mdm.commmon/src/main/java/org/talend/mdm/commmon/metadata/compare/Change.java]: If you added new property, you'd better to update the equals and hashCode method.

File: main/plugins/org.talend.mdm.commmon/src/main/java/org/talend/mdm/commmon/metadata/compare/HibernateStorageImpactAnalyzer.java [hunk 2]
```diff
@@ -160,11 +165,12 @@ protected void analyzeModifyChange(Compare.DiffResults diffResult, Map<Impact, L
                         if (!previous.isMandatory() && current.isMandatory()) {
                             // Won't be able to change constraint
                             String defaultValue = ((FieldMetadata) current).getData(MetadataRepository.DEFAULT_VALUE);
-                            if (!modifyAction.isHasNullValue()) {
+                            boolean isHasNullValue = MapUtils.getBooleanValue(modifyAction.getData(), Change.HAS_NULL_VALUE);
```
Comment (lpw82gf) [inline main/plugins/org.talend.mdm.commmon/src/main/java/org/talend/mdm/commmon/metadata/compare/HibernateStorageImpactAnalyzer.java]: `isHasNullValue` rename to `hasNullValue`

File: main/plugins/org.talend.mdm.commmon/src/main/java/org/talend/mdm/commmon/metadata/compare/HibernateStorageImpactAnalyzer.java [hunk 3]
```diff
@@ -121,12 +122,16 @@ protected void analyzeModifyChange(Compare.DiffResults diffResult, Map<Impact, L
...
-                if (element instanceof SimpleTypeFieldMetadata
-                        && MetadataUtils.getSuperConcreteType(((FieldMetadata) element).getType()).getName().equals("string")
-                        && Integer.valueOf((String) (currentLength == null ? STRING_DEFAULT_LENGTH : currentLength))
-                                .compareTo(Integer.valueOf(
-                                        (String) (previousLength == null ? STRING_DEFAULT_LENGTH : previousLength))) > 0) {
-                    impactSort.get(Impact.LOW).add(modifyAction);
+                int newLength = Integer.parseInt((currentLength == null ? STRING_DEFAULT_LENGTH : (String) currentLength));
+                int oldLength = Integer.parseInt((previousLength == null ? STRING_DEFAULT_LENGTH : (String) previousLength));
+
+                String fieldType = MetadataUtils.getSuperConcreteType(((FieldMetadata) element).getType()).getName();
```
Comment (VictorLi-talendbj) [inline main/plugins/org.talend.mdm.commmon/src/main/java/org/talend/mdm/commmon/metadata/compare/HibernateStorageImpactAnalyzer.java]: `fieldName`?

File: main/plugins/org.talend.mdm.commmon/src/main/java/org/talend/mdm/commmon/metadata/compare/HibernateStorageImpactAnalyzer.java [hunk 4]
```diff
@@ -23,7 +24,7 @@
 
 public class HibernateStorageImpactAnalyzer implements ImpactAnalyzer {
 
-    protected final String STRING_DEFAULT_LENGTH = "255"; //$NON-NLS-1$
+    public static final String STRING_DEFAULT_LENGTH = "255"; //$NON-NLS-1$
```
Comment (YohannTalend) [inline main/plugins/org.talend.mdm.commmon/src/main/java/org/talend/mdm/commmon/metadata/compare/HibernateStorageImpactAnalyzer.java]: https://docs.oracle.com/javase/tutorial/java/javaOO/accesscontrol.html
Has an interest to change the scope of the constant ?

**Pull request merged**